Color balance module : add the ASC CDL mode

aurelienpierre opened the pull request:
As Blender and many video editors, add a color balance mode based on the ASC color decision list https://en.wikipedia.org/wiki/Color_Decision_List

This allows to color-grade in linear ProPhoto RGB space, whereas the previous version was in gamma-corrected sRGB and led to saturation issues. 

The previous mode is retained as an option. The legacy params are imported. The OpenCL and SSE2 versions are provided and working (as far as I tested).

aurelienpierre commented:
Warning : with DNG files, changing the main power value gives either a green or a magenta color cast. Tested with NEF, RAF, ARW and CR files, it works as expected.

aurelienpierre commented:
DNG files work now as expected. I have added the fulcrum contrast and master saturation options.

In the pure C variant, just enabling the module gives en exposure boost. Something is wrong with the Lab <-> ProphotoRGB conversion in C, I can't find what. (I have checked the conversion matrices from 3 sources). Using sRGB fixes the problem for the pure C version (I suspect the gamma conversion does the trick).

SSE and OpenCL versions behave as expected with Prophoto RGB conversion and everything.

TurboGit commented:
Good job Aurélien. I'm probably not the right person to merge this PR, but it looks good to me and works as expected. I expect houz or hanatos to have a deeper look. Thanks.

aurelienpierre commented:
Thanks Pascal ! If you compile my master branch, you will get it along with the Log unbreak profile, and both together work really well to recover high dynamic range images.

aurelienpierre commented:
I have added HSL sliders in addition to the RGB ones, as I requested here https://redmine.darktable.org/issues/11959.

This doesn't change the data structure (parameters are still RGB). Editing RGB sliders updates HSL ones, and the other way around.

I still have a problem of slight magenta shift with the SSE ProphotoRGB conversion, OpenCL and pure C behave as expected.

I have rewritten SSE color conversion function in algebraic notation instead of SSE intrinsics to improve readability.

edgardoh commented:
I've playing with this as well, very nice! Results are very good and is very easy to get them.
Maybe add a combo with lift/gamma/gain/(all) so it doesn't take so much space on the UI?
Also, is it possible to have a new group for the entire range (not only shadows/highlights/midtones)? Sometimes I like to create a color cast and then play with each range in particular.

aurelienpierre commented:
> Maybe add a combo with lift/gamma/gain/(all) so it doesn't take so much space on the UI?

You mean on the text label ?

> Also, is it possible to have a new group for the entire range (not only shadows/highlights/midtones)?

Like a master hue setting ?

aurelienpierre commented:
> I mean to add a combo box with those entries that show/hide each group (or all of them if (all) is selected)

I still don't understand :grin: You want to selectively enable/disable the RGB set of sliders and the HSL ones ? 

> Exactly. It can be added to the combo box above, maybe as a default.

The problem with that is it would change the maths behind and lose the compatibility with the ACL Color Decision List standard. Technically-speaking, the slope is sort of the master (it's the linear factor), and the first parameter you should set in your workflow. Then, you can come back a bit on it by setting the offset, which, again applies to the whole range but, numerically, weighs more on the low-lights. And as the last step, you fix the power. That's why the mode is called slope/offset/power, in the order you are supposed to adjust the settings.

edgardoh commented:
> I mean to add a combo box with those entries that show/hide each group (or
> all of them if (all) is selected)
>
> I still don't understand 😁 You want to selectively enable/disable the
> RGB set of sliders and the HSL ones ?
>
Not disable but hide it, kind of what the channel mixer does, but with the
option of displaying them all. But if you're not going to add the master
hue, I can live with all the sliders visible, they fit in my screen anyway.


>
>

aurelienpierre commented:
I have added neutralization color-pickers on each hue slider. The idea is to select a color that should be neutral (grey) and input the right parameters in the color balance to invert its current hue and neutralize it. The saturation is not always on point though.

TurboGit commented:
You need to remove the rawspeed submodule from here.
Also there is a small conflict to be resolved.

I think also that it should be better to rebase all this and maybe even squash all the commits together at this stage. No need to see all the "fix" steps now.

aurelienpierre commented:
Isn't rawspeed listed in the gitignore that it keeps showing up in the commits ?

I will fix that, maybe add color-pickers for lift and gain factors, and I think we wil be good.

TurboGit commented:
Ok, let me know when ready. We are approaching the code freeze I suppose. So if we want this in the next release (and I think it would be really nice) I need to review and merge soon. Thanks.

aurelienpierre commented:
I think it's clean ?

So, last changes:

1. I added a load of comments to explain the maths happening in the optimizers
1. I added the optimizer for luma distribution, as promised
1. these optimizers should be a tad more robust in difficult cases (I added constraints)
1. I removed the RGB sliders bound (for luma normalization) for the legacy lift/gamma/gain, so people can continue to use it as before. It's still there for the 
1. I cherry-picked the common lib `colorspaces_inline_conversions.h` that I optimized, corrected, and made more readable by replacing the SS2 special functions by algebraic operations (following the advice of one Rawtherapee dev, because it seems algebraic operations can be more optimized by the compilator).
1. I added the `SIMD()` parameter for openmp loops
1. I added some more info to the labels.

Screenshots:

HSL sliders:
![capture du 2018-10-13 01-34-00](https://user-images.githubusercontent.com/2779157/46901790-2ebfdc80-ce88-11e8-9624-90af6e0ae879.png)

RGB sliders:
![capture du 2018-10-13 01-34-14](https://user-images.githubusercontent.com/2779157/46901791-3089a000-ce88-11e8-9d9b-4ea949f3fd60.png)

Both sliders:
![capture du 2018-10-13 01-34-31](https://user-images.githubusercontent.com/2779157/46901792-367f8100-ce88-11e8-934b-b2460ff30cfe.png)

Directions to test:

- Always set-up the slope, then lift, then gamma in this order.
- Always set-up the luma before the color.
- The factor color-pickers take the max luma in the zone for the slope, the min luma in the zone for the offset, the average luma for the power. Choose your zones accordingly. Once you have set 3 samples (slope/offset/power), you can trigger the "optimize luma" optimizer. It will fit the histogram inside the whole dynamic range [0-100 %] and normalize the average luma to 50 %, similarly to the levels but in RGB.
- The hue color-pickers aim at reverting the color of the samples you give them. They take the average color. Once you have set the 3 samples (slope/offset/gamma), you can trigger the "neutralize color" optimizer. The samples should be colors supposed to be neutral in real life in low/mid/high-lights. That's usefull when you have several white balances due to several light sources in the same picture.
- Set the master saturation and contrast in last
- The grey fulcrum is the luma value that is non-affected by the contrast adjustement. In portrait, use the color-picker to select the whole face and get its average value. Then, the contrast you will add will enhance the shape and features. 18 % Luma = 50 % L(ab) so you should always get 18 % if the gamma/power factor was set properly.

Note that the color-pickers are designed for the slope/offset/power version, as well as the HSL -> RGB conversion and the RGB luma normalization. But they still work fairly for lift/gamma/gain.

You will notice that if you try RGB -> HSL -> RGB' conversions on the sliders, for primaries, you will always have an error : RGB = RGB' ± 0.005. That's expected, due to the conversion of color shiftts between a cylindric and a cartesian vectorial space. However, ping me if you get errors >= 0.01, it shouldn't happen.

Enjoy !

TurboGit commented:
One issue found:
- open module, select "both" for "color control sliders"
- click on the color picker IN the color picker module on the left
- the color balance module is reset to "HSL" mode only instead of both

TurboGit commented:
Another one:
- open a new picture
- activate color balance
- let all default value
- click on the picker for the factor in offset
- draw an area
=> nothing happens

Is that expected? Maybe a wrong procedure on my side?

aurelienpierre commented:
The first issue is a real issue. For the second, you need to click once on the color picker, draw the area, then click again on the color picker icon to validate the color.

aurelienpierre commented:
The first issue is solved @TurboGit

TurboGit commented:
Hi Aurélien, testing under gdb I get many crashes, one example:

`#0  0x00007fffb57c0239 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#1  0x00007fffb57c3df7 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#2  0x00007fffb57bbe5a in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#3  0x00007fffb57bbf7f in ?? ()
   from /usr/lib/x86_64-linux-gnu/libnvidia-opencl.so.1
#4  0x00007ffff7dc3523 in dt_opencl_enqueue_kernel_2d_with_local (
    dev=dev@entry=0, kernel=59, sizes=0x7fffb77e64d0, local=local@entry=0x0)
    at /home/obry/dev/OpenSource/builds/darktable/build/src/src/common/opencl.c:1931
#5  0x00007ffff7dc3557 in dt_opencl_enqueue_kernel_2d (dev=dev@entry=0, 
    kernel=<optimized out>, sizes=<optimized out>)
    at /home/obry/dev/OpenSource/builds/darktable/build/src/src/common/opencl.c:1915
#6  0x00007fffb40e4edb in process_cl (self=<optimized out>, 
    piece=<optimized out>, dev_in=<optimized out>, dev_out=<optimized out>, 
    roi_in=<optimized out>, roi_out=<optimized out>)
    at /home/obry/dev/OpenSource/builds/darktable/build/src/src/iop/colorbalance.c:525
#7  0x00007ffff7e3e251 in dt_dev_pixelpipe_process_rec (
    pipe=pipe@entry=0x5555573d9210, dev=dev@entry=0x555557407890, 
    output=output@entry=0x7fffb77e7468, 
    cl_mem_output=cl_mem_output@entry=0x7fffb77e7470, 
    out_format=out_format@entry=0x7fffb77e7478, 
    roi_out=roi_out@entry=0x7fffb77e6990, modules=0x5555572eb9a0, pos=31, 
    pieces=<optimized out>, pieces=<optimized out>)
    at /home/obry/dev/OpenSource/builds/darktable/build/src/src/develop/pixelpipe_hb.c:958
#8  0x00007ffff7e3b68f in dt_dev_pixelpipe_process_rec (
    pipe=pipe@entry=0x5555573d9210, dev=dev@entry=0x555557407890, 
    output=output@entry=0x7fffb77e7468, 
    cl_mem_output=cl_mem_output@entry=0x7fffb77e7470, 
    out_format=out_format@entry=0x7fffb77e7478, 
    roi_out=roi_out@entry=0x7fffb77e6d40, modules=0x5555572ec840, pos=32, 
    pieces=<optimized out>, pieces=<optimized out>)
    at /home/obry/dev/OpenSource/builds/darktable/build/src/src/develop/pixelpipe_hb.c:599
`

TurboGit commented:
Sorry I missed the more important, it reproduces only when using the optimizer. Set the 3 factors, click optimize luma got to lighttable and quit darktable. In most cases I get a crash.

aurelienpierre commented:
It should be fixed by now, it was a problem of unproperly allocated and freed private data structure. C is not nice.

aurelienpierre commented:
Actually, it still breaks and I have no clue. The faulty lines are:

https://github.com/darktable-org/darktable/blob/68ca31eb79edb3a6324e1ad19fe5b7516022704c/src/iop/colorbalance.c#L1470-L1473

They occur in every color-picker callback and are intended to store the RGB values in a private data structure to be fetched later by the optimizer.

Commenting these lines fixes the problem, and I absolutely don't see why. It doesn't seem to be an out-of-bound error while indexing the arrays.

TurboGit commented on src/iop/colorbalance.c:
```diff
@@ -548,19 +1369,625 @@ static gboolean dt_iop_area_draw(GtkWidget *widget, cairo_t *cr, dt_iop_module_t
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  
```
In all color picker callback routines you want to activate the module if not already the case:
`if(self->off) gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(self->off), 1);
`

TurboGit commented on src/iop/colorbalance.c:
```diff
@@ -548,19 +1369,625 @@ static gboolean dt_iop_area_draw(GtkWidget *widget, cairo_t *cr, dt_iop_module_t
+  dt_iop_request_focus(self);
+  dt_control_queue_redraw();
+
+  if(self->request_color_pick == DT_REQUEST_COLORPICK_MODULE)
```
The control flow is a bit complex and seems to be equivalent to:

```
if(self->request_color_pick == DT_REQUEST_COLORPICK_OFF)
{
    dt_iop_request_focus(self);
    self->request_color_pick = DT_REQUEST_COLORPICK_MODULE;
}
else
{
    dt_iop_colorbalance_params_t *p = (dt_iop_colorbalance_params_t *)self->params;
    dt_iop_colorbalance_gui_data_t *g = (dt_iop_colorbalance_gui_data_t *)self->gui_data;

    dt_dev_reprocess_all(self->dev);

    ...
}
```

Again, this must be done in all color picker related callbacks.

TurboGit commented on src/iop/colorbalance.c:
```diff
@@ -548,19 +1369,625 @@ static gboolean dt_iop_area_draw(GtkWidget *widget, cairo_t *cr, dt_iop_module_t
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  
+  dt_iop_request_focus(self);
```
I'm not sure those 2 calls are really necessary.

TurboGit replied:
Sorry I'm speaking of dt_iop_request_focus() and dt_control_queue_redraw();

aurelienpierre replied:
`dt_control_queue_redraw()` seems mandatory to draw the frame of the color picker

TurboGit commented on src/iop/colorbalance.c:
```diff
@@ -548,19 +1369,625 @@ static gboolean dt_iop_area_draw(GtkWidget *widget, cairo_t *cr, dt_iop_module_t
+    
+    // Save the patch color for the optimization
+    dt_iop_colorbalance_data_t *d = (dt_iop_colorbalance_data_t *)self->data;
+    *d->color_patches_lift = *RGB;
```
4 floats
`float color_patches_lift[CHANNEL_SIZE];
`
3 floats:

`float RGB[3] = { 0.0f };
`

So not sure the "*left = *right" is correct. There may be some other wrong copy like this.

aurelienpierre replied:
` float color_patches_lift[CHANNEL_SIZE];`

if enum begins @ 0, `CHANNEL_SIZE` == 3, isn't it ?

TurboGit replied:
Of course not :) C is stupid. That's a quite old language at this stage and with lot of pitfall for us to kill ourselves:)

TurboGit commented on src/iop/colorbalance.c:
```diff
@@ -1608,7 +1610,7 @@ static void gain_neutralize_callback(GtkWidget *button, gpointer user_data)
     // Save the patch color for the optimization
     dt_iop_colorbalance_data_t *d = (dt_iop_colorbalance_data_t *)self->data;
-    *d->color_patches_gain = *RGB;
+    for (int c = 0; c < 3; c++) d->color_patches_gain[c] = RGB[c];
```
Is that still correct?

I mean you're mapping RGB to  CHANNEL_FACTOR, CHANNEL_RED, CHANNEL_GREEN.

Maybe: d->color_patches_gain[c+1] = RGB[c];

But this means CHANNEL_FACTOR is not set.

aurelienpierre replied:
Actually the code only uses the first 3 cells of the array, from 0, so the tags name don't matter much.

Line 1671: 

```
  for (int c = 0; c < 3; ++c) 
  {
    samples_lift[c] = CDL(d->color_patches_lift[c], p->gain[CHANNEL_FACTOR], p->lift[CHANNEL_FACTOR] - 1.0f, 1.0f / p->gamma[CHANNEL_FACTOR]);
    samples_gamma[c] = CDL(d->color_patches_gamma[c], p->gain[CHANNEL_FACTOR], p->lift[CHANNEL_FACTOR] - 1.0f, 1.0f / p->gamma[CHANNEL_FACTOR]);
    samples_gain[c] = CDL(d->color_patches_gain[c], p->gain[CHANNEL_FACTOR], p->lift[CHANNEL_FACTOR] - 1.0f, 1.0f / p->gamma[CHANNEL_FACTOR]);
  }
```

TurboGit commented:
Seems like the crash is fixed. Cannot reproduce, same on your side now?

aurelienpierre commented:
still outputs, when closing :
```
Thread 1 "darktable" received signal SIGSEGV, Segmentation fault.
(gdb) bt full
#0  0x00007ffff3e8234c in g_module_close () at /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0
#1  0x00007ffff7a7e1af in dt_view_unload_module (view=0x555557817470)
    at /home/aurelien/Documents/Programmes/darktable/src/views/view.c:228
        iter = 0x555557772580
#2  0x00007ffff7a7e1af in dt_view_manager_cleanup (vm=<optimized out>)
    at /home/aurelien/Documents/Programmes/darktable/src/views/view.c:101
        iter = 0x555557772580
#3  0x00007ffff796be61 in dt_cleanup ()
    at /home/aurelien/Documents/Programmes/darktable/src/common/darktable.c:1066
        init_gui = 1
#4  0x00007ffff7a55b41 in dt_gui_gtk_run (gui=<optimized out>)
    at /home/aurelien/Documents/Programmes/darktable/src/gui/gtk.c:1201
        widget = <optimized out>
        allocation = {x = 362, y = 12, width = 401, height = 601}
#5  0x0000555555554830 in main (argc=<optimized out>, argv=<optimized out>)
    at /home/aurelien/Documents/Programmes/darktable/src/main.c:83
```

TurboGit commented:
And related to OpenCL, I cannot reproduce with standard code path.

TurboGit commented:
And when I replace colorbalance_cdl() implementation with a simple call to colorbalance (in extended.cl) I don't have the crash (the image is not correct of course). So it looks like the issue is in the CL code in colorbalance_cdl(). Now I'm far from expert in OpenCL :(

aurelienpierre commented:
I don't understand : what crash do you get with colorbalance_cdl() ? Both versions work with no error on my side, with the last commit.

TurboGit commented:
When I'm leaving dt I get segmentation violation ini different part of the application (not always the same). This seems to indicate that there is some memory corruption somewhere.

aurelienpierre commented:
@TurboGit I have commented out the faulty lines that store the current color into the private data structure, like that:

```
#ifdef OPTIM
    dt_iop_colorbalance_data_t *d = (dt_iop_colorbalance_data_t *)self->data;
    d->luma_patches[GAIN] = XYZ[1];
    d->luma_patches_flags[GAIN] = 1;
#endif
```

Can you confirm everything works that way so that we ensure these are the true offenders ?

aurelienpierre commented:
I think I have found it.

The private data structure `(dt_iop_colorbalance_data_t *)`  is stored in `(dt_dev_pixelpipe_t *)pipe->data`. Here, I'm writing in `(dt_iop_module_t *)self->data` which is something else, hence the memory corruption.

But thanks to the barely-commented and non-documented sourcecode of dt, I have not found how to access the pixelpipe from the module structure.

TurboGit commented:
Ok, no more crashes!

And indeed valgrind (I run a session last night, very very slow under valgrind!) told me that d->luma_patches_flags[GAIN] = 1; was writing 4 bytes out of bound!

So you've found the offenders :)

aurelienpierre commented:
Ok, now I store the data in `dt_iop_colorbalance_gui_data_t` so it should be clean !!! :smile:

TurboGit commented:
Yes, all good now! Thanks.

edgardoh commented:
May I point that this can break backwards compatibility?

https://github.com/darktable-org/darktable/pull/1734/files#diff-66d4f98e149ad915383f3c1cd14d7a46R213


El mar., 16 oct. 2018 a las 10:45, Pascal Obry (<notifications@github.com>)
escribió:

> Merged #1734 <https://github.com/darktable-org/darktable/pull/1734> into
> master.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/1734#event-1906855498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APq1pT90fhLTUPbWT9MvH1tFqK6GLhEiks5uleL0gaJpZM4W1fED>
> .
>

TurboGit commented on src/common/colorspaces_inline_conversions.h:
```diff
@@ -181,12 +210,12 @@ static inline float lab_f_inv(const float x)
 static inline void dt_Lab_to_XYZ(const float *Lab, float *XYZ)
 {
-  const float d50[3] = { 0.9642, 1.0, 0.8249 };
+  const float d50[3] = { 0.96422f, 1.0f, 0.8249f };
```
I did not spot that, there is a 2 added at the end of the first value!

aurelienpierre replied:
yes, it is a typo. Thanks @edgardoh !

TurboGit replied:
I have reverted this part now.

aurelienpierre replied:
Merci beaucoup pour ton aide Pascal !
